### ISSUE TITLE
Instrument-track measure ticks sometimes off by 1px

### ISSUE THREAD

**Issue by noahb01** [LMMS/lmms#4155]
In the song-editor, the ticks on each measure for the automation track are sometimes shifted forward by 1px.

**Reproduce:**
Open new project.
Make an instrument-track and an automation-track.
Put some random stuff into the tracks, about 8-16 measures is good.
Look at the ticks on the tracks, some will not line up!

**Comment by zonkmachine** [comment]
It looks like it's the Instrument track that is a bit off. The Automation and BBTracks align with the time line.
![ticksalign](https://user-images.githubusercontent.com/6368949/35791101-124a1d60-0a46-11e8-803e-a0992ab4f43a.png)

**Comment by noahb01** [comment]
![screenshot from 2018-02-05 09-50-16](https://user-images.githubusercontent.com/22487040/35810515-01535350-0a5a-11e8-9a0c-ea634b84d5d2.png)

Interestingly on my system it appears as if the ticks "fall behind" on the Instrument tracks. The first few ticks line up but get worse as the track goes on.

Haven't looked at the code but I assume the ticks are drawn like this: Draw tick, move forward by X pixels, draw tick, ...

If this is the case then it appears that the value "X" being added is off by a very small amount.

**Comment by noahb01** [comment]
BTW, I'm using LMMS 1.2.0-rc5, Qt 5.9.2, GCC 4.8.4 AppImage build.

**Comment by zonkmachine** [comment]
It happened in commit 8c2ebf287d9de57466ffa8b1f4bda184c171dde8

```
8c2ebf287d9de57466ffa8b1f4bda184c171dde8 is the first bad commit
commit 8c2ebf287d9de57466ffa8b1f4bda184c171dde8
Author: Dave French <dave.french3@googlemail.com>
Date:   Fri Dec 26 12:33:00 2014 +0000

    Proposed fix for 1255 Segment/clip not showing whole last note in Song Editor
```

Changing https://github.com/LMMS/lmms/blob/stable-1.2/src/tracks/Pattern.cpp#L887 to
`( width() - TCO_BORDER_WIDTH )` seem to help a bit. The ticks still look a bit off but it's the same for all tracks. @noahb01 If you want to poke this one a bit that's fine with me. I just get the feeling this could be a frustrating job of pushing pixels around. You've been warned.

**Comment by PhysSong** [comment]
@zonkmachine `( width() - 1 )` looks better. See this:
https://github.com/LMMS/lmms/blob/c3b47678368d1a617c3d58b062901ee4b732c4c9/src/core/Track.cpp#L458-L461

And I think `x_base + static_cast<int>( ppt * t ) - 1` should be `x_base + static_cast<int>( ppt * t ) - 2` or `static_cast<int>( ppt * t )` [here](https://github.com/LMMS/lmms/blob/c3b47678368d1a617c3d58b062901ee4b732c4c9/src/tracks/Pattern.cpp#L1043-L1049) for consistency with [BB track](https://github.com/LMMS/lmms/blob/c3b47678368d1a617c3d58b062901ee4b732c4c9/src/tracks/BBTrack.cpp#L260-L262) and/or [automation track](https://github.com/LMMS/lmms/blob/c3b47678368d1a617c3d58b062901ee4b732c4c9/src/gui/AutomationPatternView.cpp#L369). After applying two fix, everything should be fixed.

FYI, reverting 8c2ebf2 doesn't seem to cause any regressions(#1255 or other things) on `master`.

**Comment by noahb01** [comment]
Only the `( width() - 1 )` change was required to fix the issue for me. @PhysSong what is the effect of the core/Track.cpp change? Is it only for consistency or does it have an effect on the output?

![screenshot from 2018-02-06 20-12-06](https://user-images.githubusercontent.com/22487040/35893197-e83aea5a-0b7a-11e8-8314-ae3f089bf89b.png)

**Comment by noahb01** [comment]
If this is all that needs to be fixed, should I make a pull request? Also, should I use the branch `master` or `stable-1.2`?

**Comment by noahb01** [comment]
Also it should be noted that the code for this in `master` and `stable-1.2` is quite different...

**Comment by musikBear** [comment]
imo a 1.3+ 'issue'
noone should use time on this now -just saying
Currently there are two bug-categories: *Bug* & *Critical*. imo a category *Minor* could be used for things like this.

**Comment by tresf** [comment]
@musikBear is not a project maintainer.  @PhysSong is.  Please leave this decision to the maintainers.

**Comment by musikBear** [comment]
@tresf How about not going after the man, but the ball?

**Comment by tresf** [comment]
> so what about making an argument 

@musikBear I'm not a technical reviewer for this bug so I am leaving the decision up to those that understand the codebase well enough to milestone it.

> I make a totally legit remark and suggestion, that will be beneficial, 

It's not beneficial.  I've received multiple complaints from developers that read a comment of yours and believe it reflects on the project.  I'm letting @noahb01 know that your thoughts to no necessarily reflect the project as a whole.  The team spends a lot of time supporting the users as well as the developers of the project and that can cause a delay when someone like @noahb01 asks a simple question: "What branch to target?".   Having worked with both @PhysSong and @zonkmachine on hundreds of issues, I'm fairly confident they will come back with a well thought reply that answers the question directly.

What I'm not confident about is how your comments will be received, so I'm offering insight for @noahb01 as well as future bug submitters.

So when you say "go after the ball", I have to reply with, I do.  I'm working on some of my own tasks with the project currently.  I'd be happy to offer supporting evidence of this offline if you're interested.  You know how to contact me. :)

**Comment by zonkmachine** [comment]
> Also it should be noted that the code for this in master and stable-1.2 is quite different...

As @PhysSong dived deeper into this he should probably be the one to answer. My first impression though is that it would be OK to fix `stable-1.2` with a one-liner and leave the rest for `master`. With 'the rest' here being the pattern shooting over to the right.

![patternglitch](https://user-images.githubusercontent.com/6368949/36010202-88888856-0d50-11e8-9c56-2b1893c0cc57.png)


I don't currently build `master` so I can't chip in there. 

.
`<ot>`

> I make a totally legit remark and suggestion, that will be beneficial,

This discussion should probably go somewhere else but I don't find it beneficial to have people not involved in the coding trying to organize the workflow of it. Please leave the issue labels and milestones to the coders.

**Comment by noahb01** [comment]
I'm not closing this as the issue is still present in `master`, but it has been fixed in `stable-1.2` in PR #4171.

**Comment by zonkmachine** [comment]
> @zonkmachine ( width() - 1 ) looks better. See this:

I didn't get that, sorry. I tried #4171 now and it works. I think we can merge it now for rc6 and you can refactor the code later if needed.

**Comment by Umcaruje** [comment]
Closed in #4171
